ambry-router/src/test/java/com.github.ambry.router/PutOperationTest.java

```diff
@@ -192,6 +192,44 @@ public void testSendIncomplete() throws Exception {
     Assert.assertTrue("Operation should be complete at this time", op.isOperationComplete());
   }
 
+  /**
+   * Test the Errors {@link RouterErrorCode} received by Put Operation. The operation exception is set
+   * based on the priority of these errors.
+   * @throws Exception
+   */
+  @Test
+  public void testSetOperationExceptionAndComplete() throws Exception {
+    int numChunks = NonBlockingRouter.MAX_IN_MEM_CHUNKS + 1;
+    BlobProperties blobProperties =
+        new BlobProperties(-1, "serviceId", "memberId", "contentType", false, Utils.Infinite_Time,
+            Utils.getRandomShort(TestUtils.RANDOM), Utils.getRandomShort(TestUtils.RANDOM), false);
+    byte[] userMetadata = new byte[10];
+    byte[] content = new byte[chunkSize * numChunks];
+    random.nextBytes(content);
+    ReadableStreamChannel channel = new ByteBufferReadableStreamChannel(ByteBuffer.wrap(content));
+    FutureResult<String> future = new FutureResult<>();
+    MockNetworkClient mockNetworkClient = new MockNetworkClient();
+    PutOperation op =
+        new PutOperation(routerConfig, routerMetrics, mockClusterMap, responseHandler, new LoggingNotificationSystem(),
+            userMetadata, channel, future, null,
+            new RouterCallback(mockNetworkClient, new ArrayList<BackgroundDeleteRequest>()), null, null, null, null,
+            time, blobProperties);
+    RouterErrorCode[] routerErrorCodes = new RouterErrorCode[4];
+    routerErrorCodes[0] = RouterErrorCode.OperationTimedOut;
+    routerErrorCodes[1] = RouterErrorCode.UnexpectedInternalError;
+    routerErrorCodes[2] = RouterErrorCode.AmbryUnavailable;
+    routerErrorCodes[3] = RouterErrorCode.InsufficientCapacity;
+    for (int i = 0; i < routerErrorCodes.length; ++i) {
+      op.setOperationExceptionAndComplete(new RouterException("RouterError", routerErrorCodes[i]));
+      Assert.assertEquals(((RouterException) op.getOperationException()).getErrorCode(), routerErrorCodes[i]);
+    }
+    for (int i = routerErrorCodes.length - 1; i >= 0; --i) {
+      op.setOperationExceptionAndComplete(new RouterException("RouterError", routerErrorCodes[i]));
+      Assert.assertEquals(((RouterException) op.getOperationException()).getErrorCode(),
+          routerErrorCodes[routerErrorCodes.length - 1]);
+    }
```

Comment:
also test the other parts of that function (setting a non `RouterException`)

Reply:
Got it. Will add non RouterException in tests.

Reply:
non RouterException test added.